### PR TITLE
[flutter_tools] Allow the tool to suppress compilation errors.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

suppress compilation errors on startup to make the most egregious parts of https://github.com/flutter/flutter/issues/58337 better

Specifically:

> what I think is happening here is that the tool initializes the frontend_server for the hot reload cycle concurrent with whatever native build step is running (Gradle/Xcode). This speeds up the initialization of flutter run a decent amount, about 5 - 10 seconds depending on how "dirty" the build is. Something I didn't think about when doing this was errors! Now we're getting them twice whenever there is a problem with Dart code.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
This also removes some of the extraneous output and newlines to simplify things a bit. For example, a compile error on startup (android) now looks like:

```
Launching lib\main.dart on Pixel 4 in debug mode...
lib/main.dart:12:11: Error: Expected ',' before this.
          key: Key('title'),
          ^^^
lib/main.dart:11:13: Error: Too many positional arguments: 1 allowed, but 2 found.
Try removing the extra positional arguments.
        Text('Hello, world!',2
            ^
../../packages/flutter/lib/src/widgets/text.dart:282:9: Context: Found this candidate, but the arguments don't match.
  const Text(
        ^^^^


FAILURE: Build failed with an exception.

* Where:
Script 'C:\Users\Jonah\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 896

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command 'C:\Users\Jonah\flutter\bin\flutter.bat'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
Running Gradle task 'assembleDebug'...
Running Gradle task 'assembleDebug'... Done                         4.7s
Exception: Gradle task assembleDebug failed with exit code 1
```

In a separate PR, we could chop off the footer (seems fairly stable) when the error is `app:compileFlutterBuildDebug` or another flutter-related task and replace it with a better error message. (i.e. verbose instead of stack-trace

File: packages/flutter_tools/lib/src/build_system/targets/dart.dart [hunk 1]
```diff
@@ -258,7 +258,7 @@ class KernelSnapshot extends Target {
       packageConfig: packageConfig,
     );
     if (output == null || output.errorCount != 0) {
-      throw Exception('Errors during snapshot creation: $output');
+      throw Exception();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/dart.dart:261]: This adds nothing

File: packages/flutter_tools/lib/src/commands/assemble.dart [hunk 2]
```diff
@@ -226,13 +226,13 @@ class AssembleCommand extends FlutterCommand {
       );
     if (!result.success) {
       for (final ExceptionMeasurement measurement in result.exceptions.values) {
-        globals.printError('Target ${measurement.target} failed: ${measurement.exception}',
-          stackTrace: measurement.fatal
-            ? measurement.stackTrace
-            : null,
-        );
+        if (measurement.fatal || globals.logger.isVerbose) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/assemble.dart:229]: Only show for verbose errors, actual bugs will get logged to stderr

File: packages/flutter_tools/lib/src/compile.dart [hunk 3]
```diff
@@ -101,44 +100,13 @@ class StdoutHandler {
...
-    }
     const String kResultPrefix = 'result ';
     if (boundaryKey == null && message.startsWith(kResultPrefix)) {
       boundaryKey = message.substring(kResultPrefix.length);
       return;
     }
-    // Invalid state, see commented issue below for more information.
-    // NB: both the completeError and _badState flags are required to avoid
-    // filling the console with exceptions.
-    if (boundaryKey == null) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/compile.dart:118]: This was fixed several months ago by @aam

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart:118]: https://github.com/flutter/flutter/issues/35924

File: packages/flutter_tools/test/general.shard/resident_runner_test.dart [hunk 4]
```diff
@@ -207,6 +207,93 @@ void main() {
...
+    fakeVmServiceHost = FakeVmServiceHost(requests: <VmServiceExpectation>[
+      listViews,
+      listViews,
+    ]);
+    final MockResidentCompiler residentCompiler = MockResidentCompiler();
+    residentRunner = HotRunner(
+      <FlutterDevice>[
+        mockFlutterDevice,
+      ],
+      applicationBinary: globals.fs.file('app.apk'),
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/resident_runner_test.dart:265]: When there is an applicationBinary then there was no native build step, so we need to show the flutter run errors as normal

**Comment by jonahwilliams** [comment]
The xcode output is:

```
Running Xcode build...                                                  
                                                   
Xcode build done.                                            5.6s
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **


Xcode's output:
↳
    lib/main.dart:16:5: Error: Expected ';' after this.
        )
        ^
    lib/main.dart:17:1: Error: Unexpected token ''.

    Command PhaseScriptExecution failed with a nonzero exit code
    note: Using new build systemnote: Planning buildnote: Constructing build description

Could not build the application for the simulator.
Error launching application on iPhone 11 Pro Max.
```

Which is a bit better

File: packages/flutter_tools/lib/src/codegen.dart [hunk 5]
```diff
@@ -178,6 +178,7 @@ class CodeGeneratingResidentCompiler implements ResidentCompiler {
     List<Uri> invalidatedFiles, {
       String outputPath,
       PackageConfig packageConfig,
+      bool suppressErrors = false,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/codegen.dart:181]: Can you add a comment for when you would want to suppress errors?  Doesn't sound like something you would generally want to do.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/codegen.dart:181]: done here: https://github.com/flutter/flutter/pull/58539/files#diff-e78651ad5bd5c46811f2b76bc465220bR375

Comment (jmagman) [reply packages/flutter_tools/lib/src/codegen.dart:181]: I guess this PR is the crumb trail for the signature as to why you would want to suppress errors (so they aren't displayed twice in certain scenarios).

**Review by jmagman — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
We should consider having end-to-end tests that verify the entire text of common scenarios like error messages, startup, etc, to make sure we don't accidentally regress the actual output.

**Pull request merged**